src/app/application/services/data-source/DataSourcePool.ts

```diff
@@ -0,0 +1,80 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const create = (src: string, router: Router<Source>): CallableEventSource => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
+    }
+  })
+  return _source
+}
+const destroy = (_src: string, source: CallableEventSource) => {
+  if (source) {
+    source.close()
+  }
+}
+
+export class DataSourcePool {
+  cache: Map<string, unknown> = new Map()
```

Comment:
Worth adding a comment saying that this cache is unbounded but it's just a buffer to limit the use of loaders.

Reply:
Good call, we should talk about this fact deeper at a later date as I definitely would like to add something to address this. I'll add something in comments for now.